pkg/monitor/controller.go

```diff
@@ -0,0 +1,154 @@
+	monitor.deleteStaleAgentCRDs()
+
+	// Update controller monitoring CRD variables every 60 seconds util stopCh is closed.
+	wait.PollUntil(60*time.Second, func() (done bool, err error) {
+		controllerCRD, err = monitor.partialUpdateControllerCRD(controllerCRD)
+		if err != nil {
+			klog.Errorf("Failed to partially update controller monitoring CRD %+v: %v", controllerCRD, err)
+		}
+		return false, nil
+	}, stopCh)
+}
+
+// getControllerCRD is used to check the existence of controller monitoring CRD.
+// So when the pod restarts, it will update this monitoring CRD instead of creating a new one.
+func (monitor *controllerMonitor) getControllerCRD(crdName string) *v1beta1.AntreaControllerInfo {
+	controllerCRD, err := monitor.client.ClusterinformationV1beta1().AntreaControllerInfos().Get(crdName, metav1.GetOptions{})
+	if err != nil {
+		klog.V(2).Infof("Controller monitoring CRD named %s doesn't exist, will create one", crdName)
+		return nil
+	}
+	return controllerCRD
+}
+
+func (monitor *controllerMonitor) createControllerCRD(crdName string) (*v1beta1.AntreaControllerInfo, error) {
+	controllerCRD := new(v1beta1.AntreaControllerInfo)
+	monitor.querier.GetControllerInfo(controllerCRD, false)
+	controllerCRD.ObjectMeta.Name = crdName
+	klog.V(2).Infof("Creating controller monitoring CRD %+v", controllerCRD)
+	return monitor.client.ClusterinformationV1beta1().AntreaControllerInfos().Create(controllerCRD)
+}
+
+// updateControllerCRD updates all the fields of existing monitoring CRD.
+func (monitor *controllerMonitor) updateControllerCRD(controllerCRD *v1beta1.AntreaControllerInfo) (*v1beta1.AntreaControllerInfo, error) {
+	monitor.querier.GetControllerInfo(controllerCRD, false)
+	klog.V(2).Infof("Updating controller monitoring CRD %+v", controllerCRD)
+	return monitor.client.ClusterinformationV1beta1().AntreaControllerInfos().Update(controllerCRD)
+}
+
+// partialUpdateControllerCRD only updates the variables.
+func (monitor *controllerMonitor) partialUpdateControllerCRD(controllerCRD *v1beta1.AntreaControllerInfo) (*v1beta1.AntreaControllerInfo, error) {
+	monitor.querier.GetControllerInfo(controllerCRD, true)
+	klog.V(2).Infof("Partially updating controller monitoring CRD %+v", controllerCRD)
+	return monitor.client.ClusterinformationV1beta1().AntreaControllerInfos().Update(controllerCRD)
+}
+
+func (monitor *controllerMonitor) deleteStaleAgentCRDs() {
+	crds, err := monitor.client.ClusterinformationV1beta1().AntreaAgentInfos().List(metav1.ListOptions{})
+	if err != nil {
+		klog.Errorf("Failed to list agent monitoring CRDs: %v", err)
+		return
+	}
+	// Delete stale agent monitoring CRD based on existing nodes.
+	nodeLister := monitor.nodeInformer.Lister()
+	for _, crd := range crds.Items {
+		_, err := nodeLister.Get(crd.Name)
+		if errors.IsNotFound(err) {
+			monitor.deleteAgentCRD(crd.Name)
+		}
+	}
+}
+
+func (monitor *controllerMonitor) deleteStaleAgentCRD(old interface{}) {
+	node := old.(*v1.Node)
+	monitor.deleteAgentCRD(node.Name)
```

Comment:
Probably need to handle https://github.com/vmware-tanzu/antrea/issues/579 too

Reply:
Updated. PTAL.